src/EnergyPlus/ExhaustAirSystemManager.cc

```diff
@@ -0,0 +1,940 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// ObjexxFCL Headers
+#include <ObjexxFCL/Array.functions.hh>
+#include <ObjexxFCL/Fmath.hh>
+
+// EnergyPlus Headers
+#include <AirflowNetwork/Elements.hpp>
+#include <EnergyPlus/Autosizing/Base.hh>
+#include <EnergyPlus/BranchNodeConnections.hh>
+#include <EnergyPlus/Data/EnergyPlusData.hh>
+#include <EnergyPlus/DataContaminantBalance.hh>
+#include <EnergyPlus/DataEnvironment.hh>
+#include <EnergyPlus/DataHVACGlobals.hh>
+#include <EnergyPlus/DataHeatBalFanSys.hh>
+#include <EnergyPlus/DataHeatBalance.hh>
+#include <EnergyPlus/DataIPShortCuts.hh>
+#include <EnergyPlus/DataLoopNode.hh>
+#include <EnergyPlus/DataSizing.hh>
+#include <EnergyPlus/DataZoneEquipment.hh>
+#include <EnergyPlus/ExhaustAirSystemManager.hh>
+#include <EnergyPlus/Fans.hh>
+#include <EnergyPlus/GeneralRoutines.hh>
+#include <EnergyPlus/HVACFan.hh>
+#include <EnergyPlus/InputProcessing/InputProcessor.hh>
+#include <EnergyPlus/MixerComponent.hh>
+#include <EnergyPlus/NodeInputManager.hh>
+#include <EnergyPlus/Psychrometrics.hh>
+// #include <EnergyPlus/ReturnAirPathManager.hh> //2022-01-14: replace with exhaust system
+#include <EnergyPlus/ScheduleManager.hh>
+#include <EnergyPlus/UtilityRoutines.hh>
+// #include <EnergyPlus/ZonePlenum.hh> //2022-01-14: may not needed for this exhaust system
+
+namespace EnergyPlus {
+
+namespace ExhaustAirSystemManager {
+    // Module containing the routines dealing with the AirLoopHVAC:ExhaustSystem
+
+    // map might be aa better choice:
+    // std::map<int, int> mixerBranchMap;
+    std::map<int, int> mixerIndexMap;
+
+    bool mappingDone = false;
+
+    void SimExhaustAirSystem(EnergyPlusData &state, bool FirstHVACIteration)
+    {
+        // Obtains and Allocates Mixer related parameters from input file
+        if (state.dataExhAirSystemMrg->GetInputFlag) { // First time subroutine has been entered
+            GetExhaustAirSystemInput(state);
+            state.dataExhAirSystemMrg->GetInputFlag = false;
+        }
+
+        for (int ExhaustAirSystemNum = 1; ExhaustAirSystemNum <= state.dataZoneEquip->NumExhaustAirSystems; ++ExhaustAirSystemNum) {
+            CalcExhaustAirSystem(state, ExhaustAirSystemNum, FirstHVACIteration);
+        }
+
+        // After this, update the exhaust flows according to zone grouping:
+        UpdateZoneExhaustControl(state);
+    }
+
+    void GetExhaustAirSystemInput(EnergyPlusData &state)
+    {
+        // Locals
+        bool IsNotOK; // Flag to verify name
+        bool ErrorsFound = false;
+
+        if (allocated(state.dataZoneEquip->ExhaustAirSystem)) {
+            return;
+        }
```

Comment:
I wasn't going to mention this but if I don't this will just sit here. This "if (allocated...)" doesn't do anything if all calls look like this:

        if (state.dataExhAirSystemMrg->GetInputFlag) { // First time subroutine has been entered
            GetExhaustAirSystemInput(state);
            state.dataExhAirSystemMrg->GetInputFlag = false;
        }

also, the if (allocated) is only true if an array is actually allocated, even if size=0. So, if there are no exhaust systems this will always be false. It would be better for this to look like:

    if (!state.dataExhAirSystemMrg->GetInputFlag) return;
    state.dataExhAirSystemMrg->GetInputFlag = false;

Reply:
Thanks! I think it might be just using the first line: 
```
if (!state.dataExhAirSystemMrg->GetInputFlag) return;
// state.dataExhAirSystemMrg->GetInputFlag = false;
```
and leave the second line out. Theoretically this will only let the state.dataExhAirSystemMrg->GetInputFlag be set to false after the `GetExhaustAirSystemInput(state);` is successfully executed (based on the `SimExhaustAirSystem()` call), rather than just set the input flag false in any conditions. 

The original lines  were used because of these lines (Lines 125-127): 
```
            if (numExhaustSystems > 0) {
                state.dataZoneEquip->ExhaustAirSystem.allocate(numExhaustSystems);
            }
```
where `numExhaustSystems` is equivlently `state.dataInputProcessing->inputProcessor->epJSON.find("AirLoopHVAC:ExhaustSystem").value().size()`. It seemed that the `allocate()` only got executed when there are at >0 input blocks in the idf (or json) input file.